dbms/src/Processors/ISource.cpp

```diff
@@ -45,23 +38,25 @@ void ISource::work()
 {
     try
     {
-        current_chunk.chunk = generate();
-        if (!current_chunk.chunk)
-            finished = true;
+        if (auto chunk = generate())
+        {
+            current_chunk.chunk = std::move(*chunk);
+
+            if (current_chunk.chunk)
+                has_input = true;
+            else if (reset()) /// if returned chunk is empty, try to reset source and try again next time.
```

Comment:
It's cool that now we can return empty chunks! I think it will be useful.
But now I look at this code and think that we can do this check in higher-level class. Like
```c++
class MySource : public ISource
{
...
    auto generate() override
    {
        if (auto chunk = generateImpl()) /// Call old version of generate
            return chunk;
        if (reset())
            return Chunk();
        return {};
    }
}
```

Reply:
Decided with @KochetovNicolai to keep it as-is in this PR